llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp

```diff
@@ -65,18 +69,269 @@ static void emitResourceMetadata(Module &M, const DXILResourceMap &DRM,
+          ShaderStage != Triple::EnvironmentType::Library)
+        MDVals.append(
+            getTagValueAsMetadata(ShaderKindTag, getShaderStage(ShaderStage)));
+
+      if (ShaderStage == Triple::EnvironmentType::Compute) {
+        MDVals.emplace_back(ConstantAsMetadata::get(
+            ConstantInt::get(Type::getInt32Ty(Ctx), NumThreadsTag)));
+        std::vector<Metadata *> NumThreadVals;
+        NumThreadVals.emplace_back(ConstantAsMetadata::get(
+            ConstantInt::get(Type::getInt32Ty(Ctx), NumThreadsX)));
+        NumThreadVals.emplace_back(ConstantAsMetadata::get(
+            ConstantInt::get(Type::getInt32Ty(Ctx), NumThreadsY)));
+        NumThreadVals.emplace_back(ConstantAsMetadata::get(
+            ConstantInt::get(Type::getInt32Ty(Ctx), NumThreadsZ)));
+        MDVals.emplace_back(MDNode::get(Ctx, NumThreadVals));
+      }
+    }
+    if (MDVals.empty())
+      return nullptr;
+    return MDNode::get(Ctx, MDVals);
+  }
+};
+
+static void createEntryMD(Module &M, const uint64_t ShaderFlags,
+                          const dxil::ModuleMetadataInfo &MDAnalysisInfo) {
+  auto &Ctx = M.getContext();
+  // FIXME: generate metadata for resource.
+  MDTuple *MDResources = nullptr;
+  if (auto *NamedResources = M.getNamedMetadata("dx.resources"))
+    MDResources = dyn_cast<MDTuple>(NamedResources->getOperand(0));
+
+  std::vector<MDNode *> EntryFnMDNodes;
+  switch (MDAnalysisInfo.ShaderProfile) {
+  case Triple::EnvironmentType::Library: {
+    // Library has an entry metadata with resource table metadata and all other
+    // MDNodes as null.
+    EntryProperties EP{};
+    // FIXME: ShaderFlagsAnalysis pass needs to collect and provide ShaderFlags
+    // for each entry function. Currently, ShaderFlags value provided by
+    // ShaderFlagsAnalysis pass is created by walking *all* the function
+    // instructions of the module. Is it is correct to use this value for
+    // metadata of the empty library entry?
+    ShaderEntryMDInfo EmptyFunEntryProps(EP, Ctx, MDAnalysisInfo.ShaderProfile,
+                                         MDResources, ShaderFlags);
+    MDTuple *EmptyMDT = EmptyFunEntryProps.getAsMetadata();
+    EntryFnMDNodes.emplace_back(EmptyMDT);
+
+    for (auto EntryProp : MDAnalysisInfo.EntryPropertyVec) {
+      // FIXME: ShaderFlagsAnalysis pass needs to collect and provide
+      // ShaderFlags for each entry function. For now, assume shader flags value
+      // of entry functions being compiled for lib_* shader profile viz.,
+      // EntryPro.Entry is 0.
+      ShaderEntryMDInfo SEP(EntryProp, Ctx, MDAnalysisInfo.ShaderProfile,
+                            nullptr, 0);
+      MDTuple *MDT = SEP.getAsMetadata();
+      EntryFnMDNodes.emplace_back(MDT);
+    }
+  } break;
+  case Triple::EnvironmentType::Compute: {
+    size_t NumEntries = MDAnalysisInfo.EntryPropertyVec.size();
+    assert(NumEntries == 1 &&
+           "Compute shader: One and only one entry expected");
```

Comment:
We should probably diagnose an error if we have the wrong number of entry points, rather than crashing.

Reply:
> We should probably diagnose an error if we have the wrong number of entry points, rather than crashing.

Added `class DiagnosticinfoModuleFormat` and used it to report the error. Added a new test to verify that error is triggered.